network/retrieve/retrieve.go

```diff
@@ -0,0 +1,431 @@
+		Messages: []interface{}{
+			ChunkDelivery{},
+			RetrieveRequest{},
+		},
+	}
+
+	ErrNoPeerFound = errors.New("no peer found")
+)
+
+// Retrieval holds state and handles protocol messages for the `bzz-retrieve` protocol
+type Retrieval struct {
+	mtx      sync.Mutex
+	netStore *storage.NetStore
+	kad      *network.Kademlia
+	peers    map[enode.ID]*Peer //this protocols peers
+	spec     *protocols.Spec    //this protocol's spec
+	balance  protocols.Balance  //implements protocols.Balance, for accounting
+	prices   protocols.Prices   //implements protocols.Prices, provides prices to accounting
+
+	quit chan struct{} // termination
+}
+
+// NewRetrieval returns a new instance of the retrieval protocol handler
+func NewRetrieval(kad *network.Kademlia, ns *storage.NetStore) *Retrieval {
+	ret := &Retrieval{
+		kad:      kad,
+		peers:    make(map[enode.ID]*Peer),
+		netStore: ns,
+		quit:     make(chan struct{}),
+	}
+
+	ret.spec = spec
+
+	return ret
+}
+
+func (r *Retrieval) addPeer(p *Peer) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	r.peers[p.ID()] = p
+}
+
+func (r *Retrieval) removePeer(p *Peer) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	delete(r.peers, p.ID())
+}
+
+func (r *Retrieval) getPeer(id enode.ID) *Peer {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+
+	return r.peers[id]
+}
+
+// Run protocol function
+func (r *Retrieval) Run(p *p2p.Peer, rw p2p.MsgReadWriter) error {
+	peer := protocols.NewPeer(p, rw, r.spec)
+	bp := network.NewBzzPeer(peer)
+	sp := NewPeer(bp)
+	r.addPeer(sp)
+	defer r.removePeer(sp)
+	return peer.Run(r.HandleMsg(sp))
+}
+
+// HandleMsg handles incoming messages for a certain peer
+func (r *Retrieval) HandleMsg(p *Peer) func(context.Context, interface{}) error {
+	return func(ctx context.Context, msg interface{}) error {
+		switch msg := msg.(type) {
+		case *RetrieveRequest:
+			go r.handleRetrieveRequest(ctx, p, msg)
+		case *ChunkDelivery:
+			go r.handleChunkDelivery(ctx, p, msg)
+		}
+		return nil
+	}
+}
+
+// getOriginPo returns the originPo if the incoming Request has an Origin
+// if our node is the first node that requests this chunk, then we don't have an Origin,
+// and return -1
+// this is used only for tracing, and can probably be refactor so that we don't have to
+// iterater over Kademlia
+func (r *Retrieval) getOriginPo(req *storage.Request) int {
+	log.Trace("retrieval.getOriginPo", "req.Addr", req.Addr)
+	originPo := -1
+
+	r.kad.EachConn(req.Addr[:], 255, func(p *network.Peer, po int) bool {
+		id := p.ID()
+
+		// get po between chunk and origin
+		if req.Origin.String() == id.String() {
+			originPo = po
+			return false
+		}
+
+		return true
+	})
+
+	return originPo
+}
+
+// findPeer finds a peer we need to ask for a specific chunk from according to our kademlia
+func (r *Retrieval) findPeer(ctx context.Context, req *storage.Request) (retPeer *network.Peer, err error) {
+	log.Trace("retrieval.findPeer", "req.Addr", req.Addr)
+	osp, _ := ctx.Value("remote.fetch").(opentracing.Span)
```

Comment:
i wouldn't like to have this function fail on a debugging feature. leaving it as is.

Reply:
I guess what Fabio meant is to `handle the error`, not necessarily to fail the function on a debugging feature. We can log the error if it happens. Then again we `know` that this is `opentracing.Span` so I think it is not necessary.

Reply:
i understand. and i dont want to log anything because is a very non-critical error and i have a feeling that if we log it - we'll see a lot of log spamming in production